creusot/src/translation/fmir.rs

```diff
@@ -65,6 +64,47 @@ impl<'tcx> Place<'tcx> {
             None
         }
     }
+
+    fn into_term(self, tcx: TyCtxt<'tcx>, locals: &LocalDecls<'tcx>, span: Span) -> Term<'tcx> {
```

Comment:
I don't like this much, both because this means that we are introducing yet another translation for places (with maybe subtle semantic differences), and because the resulting code will have some duplication.

Instead, could we not add to FMir the possibility to annotate calls with a variant to check, before substituting the parameters ? In the backend, it should be easy to check the variant when translating the call.

Reply:
An alternative is to generate a separate handlers for recursive calls, with one additional precondition: the fact that the variant decreased.

Reply:
I did your second solution, which is somewhat similar to loop variants.